### DISCUSSION TITLE
Add environment parameter to next CLIs to natively managing different environments

### DISCUSSION THREAD

**Discussion by LucasLemanowicz** [vercel/next.js#31948] [Ideas]
### Describe the feature you'd like to request

I have a local dev environment, a deployed staging environment, and a deployed production environment. It would be nice to be able to manage all those environments with .env files without having to import additional packages or use conditional logic in the source code itself.

### Describe the solution you'd like

I'd like to be able to define an environment variable that tells NextJS which env file to load. For example, a sample `package.json` file would look like:

```javascript
{
  "scripts": {
     "build": "next build", // default = production
     "build:staging": "NEXT_ENV=staging next build",
     "build:production": "NEXT_ENV=production next build",
     // ...
  },
  // ...
}
```

Another (possibly even cleaner) alternative would be to pass in the environment as a parameter:
```javascript
{
  "scripts": {
     "build": "next build", // default = production
     "build:staging": "next build staging",
     "build:production": "next build production",
     // ...
  },
  // ...
}
```

### Describe alternatives you've considered

There is no clean way to do this in NextJS currently. Alternatives include addition conditional logic in the actual code, and/or installing additional dependencies like `env-cmd`. There are enough blog posts and medium articles about this that it's worth considering how to support these workflows in a first class way. Here's just a few:

* https://medium.com/@dzungnguyen179/nextjs-multiple-environment-builds-e8b2ccb11c04
* https://stackoverflow.com/questions/59462614/how-to-use-different-env-files-with-nextjs
* https://dev.to/martiuh/cool-way-to-use-staging-and-production-environment-with-next-js-28bb

**Comment by balazsorban44** [comment]
Have you seen the official docs about this? https://nextjs.org/docs/basic-features/environment-variables

you can define different `.env.*` files for different environments.

Here is an example repo as well https://github.com/vercel/next.js/tree/canary/examples%2Fenvironment-variables

**Comment by LucasLemanowicz** [comment]
Hey @balazsorban44, thanks for the links. Yes, I have read through them.

Correct me if I'm wrong, but the solution in that example repo requires anyone deploying to multiple environments to essentially write code like this:

`const foo = process.env.LOCAL_ENV_VARIABLE ?? process.env.DEVELOPMENT_ENV_VARIABLE ?? process.env.PRODUCTION_ENV_VARIABLE`

or requires using not-checked-in `.env.local` files that need to be clobbered together by a separate build system.

In my opinion that defeats the point of having different environment files in the first place. In fact it makes the code more prone to errors because the expected environments are now hidden in code. At that point you might as well have everything in one .env file.

The suggestion I am proposing would let you have several environment files:

```
.env.development
.env.production
.env.outerspace
```

and in your code you could just write `const foo = process.env.ENV_VARIABLE` depending on what parameters (if any) were passed into `next build`.

I feel like there's a disconnect between how I'm thinking about deployability and NextJS's philosophy about it. It is much easier to have the framework handle environment syntactically for us than needing us to get the environment and manage it in the app. Does my suggestion make sense?

**Reply by balazsorban44** [reply]
You don't have to use different names for the variables. Depending on the environment/step you are in, the correct file will be used automatically.

Eg.:

```js
const foo = process.env.ENV_VARIABLE
```

Where `ENV_VARIABLE` will be **any** of the values of the same name in the different `.env.*` files.

**Reply by LucasLemanowicz** [reply]
Hmm ok, I'm not sure I fully understand the actual implementation.

Let's say I have two environments that I deploy my application to: `alicescupcakes` and `bobsdonuts`. I would have two env files, `.env.alicescupcakes` and `.env.bobsdonuts`. How do I tell nextjs what environment I'm in?
